### PR TITLE
docs: add explanations for how to manage service dependencies

### PR DESCRIPTION
**Pull request by IronCore864**
A how-to guide on managing service dependencies.

Preview: https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/

Note that this PR include changes from https://github.com/canonical/pebble/pull/438; to only review the changes on the service dependencies how-to guide, see [here](https://github.com/IronCore864/pebble/pull/2).

### CODE REVIEW

**Comment by benhoyt** [comment]
Note that I've fixed and updated the wording in the README "service dependencies" section -- we may need to update the content in the docs in parallel (till the docs site is published and the old README content is deleted).

File: docs/how-to/service-dependencies.md [hunk 1]
```diff
@@ -1,11 +1,297 @@
...
+backend-server   enabled  active   today at 10:17 CST
+database         enabled  active   today at 10:17 CST
+frontend-server  enabled  active   today at 10:17 CST
+```
+
+## Problems
+
+Apparently, the above layer configuration works but is not exactly ideal, because:
+
+- In this example, the three components kind of rely on each other, without anyone, our whole web app won't function properly.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:112]:
```suggestion
- In this example, the three components rely on each other. Without any one of them, our whole web app won't function properly.
```

File: docs/how-to/service-dependencies.md [hunk 2]
```diff
@@ -1,11 +1,297 @@
...
+- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend starts, it fails to communicate with the database, so the whole web app won't function properly.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:113]:
```suggestion
- If the database isn't started (or can't be started for some reason), it doesn't make sense to start the backend that depends on it. Even if the backend started, it would fail to communicate with the database, so the whole web app won't function properly.
```

File: docs/how-to/service-dependencies.md [hunk 3]
```diff
@@ -1,11 +1,297 @@
...
+- The same is true for the frontend: if the backend isn't started (fails to start, can't be started, etc.), it makes no sense to start the frontend alone.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:114]:
```suggestion
- The same is true for the frontend: if the backend isn't started (if it fails to start or can't be started), it makes no sense to start the frontend alone.
```

File: docs/how-to/service-dependencies.md [hunk 4]
```diff
@@ -1,11 +1,297 @@
...
+
+For example, when the DB fails to start, other services are still up and running:
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:116]:
```suggestion
The following example shows the database failing to start, but the database and frontend starting successfully:
```

File: docs/how-to/service-dependencies.md [hunk 5]
```diff
@@ -1,11 +1,297 @@
...
+
+```{terminal}
+   :input: pebble services
+Service          Startup  Current   Since
+backend-server   enabled  active    today at 10:20 CST
+database         enabled  inactive  -
+frontend-server  enabled  active    today at 10:20 CST
+```
+
+As we can see, although `database` is in `inactive` state, the backend/frontend are still both started, which isn't exactly useful since all they are doing is consuming resources without being able to provide any service.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:138]:
```suggestion
As we can see, although the `database` is `inactive`, the backend and frontend are both `active`. This isn't exactly useful, since all they are doing is consuming resources without being able to provide any service.
```

File: docs/how-to/service-dependencies.md [hunk 6]
```diff
@@ -1,11 +1,297 @@
...
+
+Wouldn't it be nice to make these services depend on each other, and when a dependency fails to start, put other services that rely on it on hold, instead of starting them anyway?
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:140]: This is quite a long pile-up of phrases. Maybe this?

```suggestion
Wouldn't it be nice to tell Pebble that these services depend on each other? If a dependency failed to start, Pebble could put the services that rely on it "on hold" instead of starting them. This is where service dependencies come in.
```

File: docs/how-to/service-dependencies.md [hunk 7]
```diff
@@ -1,11 +1,297 @@
...
+
+Worry no more, service dependency comes to the rescue.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:142]: I think "worry no more" and "to the rescue" is just a bit too casual / unprofessional. I've rolled it into the last sentence above.

File: docs/how-to/service-dependencies.md [hunk 8]
```diff
@@ -1,11 +1,297 @@
...
+
+## Service dependencies
+
+Pebble can take service dependencies into account when managing services: the magical `required` (see more detail in [Layer specification](../reference/layer-specification.md)) feature.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:146]: Again, I think "magical" is a bit much here. I get that you're trying to have a fun tone, but I think it strikes the wrong chord for this kind of documentation. It would be fine in a personal blog post.

Also note that it's `requires`, not `required`.

```suggestion
Pebble can take service dependencies into account when managing services: this is done with the `required` list in the [service definition](../reference/layer-specification.md).
```

File: docs/how-to/service-dependencies.md [hunk 9]
```diff
@@ -1,11 +1,297 @@
...
+
+Simply put, you can configure a list of other services in the `required` section to indicate this service requires those other services to start correctly.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:148]:
```suggestion
Simply put, you can configure a list of other services in the `requires` section to indicate this service requires those other services to start correctly.
```

File: docs/how-to/service-dependencies.md [hunk 10]
```diff
@@ -1,11 +1,297 @@
...
+
+When Pebble starts a service, it also starts the services which that service depends on (configured with `required`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:150]:
```suggestion
When Pebble starts a service, it also starts the services which that service depends on (configured with `requires`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
```

File: docs/how-to/service-dependencies.md [hunk 11]
```diff
@@ -1,11 +1,297 @@
...
 
 For example, if service `nginx` requires `logger`, `pebble start nginx` will start both `nginx` and `logger` (in an undefined order). Running `pebble stop logger` will stop both `nginx` and `logger`; however, running `pebble stop nginx` will only stop `nginx` (`nginx` depends on `logger`, not the other way around).
 
-When multiple services need to be started together, they're started in order according to the `before` and `after` configuration, waiting 1 second for each to ensure the command doesn't exit too quickly. The `before` option is a list of services that this service must start before (it may or may not `require` them). Or if it's easier to specify this ordering the other way around, `after` is a list of services that this service must start after.
+## Start order
+
+When multiple services need to be started together, they're started in order according to the `before` and `after` configuration (see more detail in [Layer specification](../reference/layer-specification.md)), waiting for 1 second for each to ensure the command doesn't exit too quickly.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:156]: I wonder if we can again avoid the parenthesis and just link "configuration" directly:

```suggestion
When multiple services need to be started together, they're started in order according to the `before` and `after` in the [layer configuration](../reference/layer-specification.md). Pebble waits 1 second after starting each service to ensure the command doesn't exit too quickly.
```

File: docs/how-to/service-dependencies.md [hunk 12]
```diff
@@ -1,11 +1,297 @@
...
-Note that currently, `before` and `after` are of limited usefulness, because Pebble only waits 1 second before moving on to start the next service, with no additional checks that the previous service is operating correctly.
+```{note}
+Currently, `before` and `after` are of limited usefulness, because Pebble only waits 1 second before moving on to start the next service, with no additional checks that the previous service is operating correctly.
 
 If the configuration of `requires`, `before`, and `after` for a service results in a cycle or "loop", an error will be returned when attempting to start or stop the service.
+```
+
+## Putting it together
+
+With the service dependencies and start order features, let's reconfigure our web app layer to:
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:168]:
```suggestion
With the service dependencies and start order features, let's reconfigure our web app layer:
```

File: docs/how-to/service-dependencies.md [hunk 13]
```diff
@@ -1,11 +1,297 @@
...
+If we try to start it now:
+
+```{terminal}
+   :input: pebble run
+2024-06-28T02:17:23.347Z [pebble] Started daemon.
+[backend-server database frontend-server]
+2024-06-28T02:17:23.353Z [pebble] POST /v1/services 2.613042ms 202
+2024-06-28T02:17:23.356Z [pebble] Service "backend-server" starting: python3 -m http.server 8081
+2024-06-28T02:17:24.363Z [pebble] Service "database" starting: python3 -m http.server 3306
+2024-06-28T02:17:25.370Z [pebble] Service "frontend-server" starting: python3 -m http.server 8080
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:208]: With the configuration above, I would expect to see database, backend-server, then frontend-server -- in that order due to the `after` configuration. That's what I get locally -- was this a copy 'n' paste error?

File: docs/how-to/service-dependencies.md [hunk 14]
```diff
@@ -1,11 +1,297 @@
...
+2024-06-28T02:17:26.385Z [pebble] Started default services with change 1.
+```
+
+And we can verify the status of the services:
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:212]:
```suggestion
You can see how `database` is started first, then `backend-server`, then `frontend-server` -- this is due to the `after` ordering. And we can verify the status of the services:
```

File: docs/how-to/service-dependencies.md [hunk 15]
```diff
@@ -1,11 +1,297 @@
...
+
+```{terminal}
+   :input: pebble services
+Service          Startup  Current  Since
+backend-server   enabled  active   today at 10:17 CST
+database         enabled  active   today at 10:17 CST
+frontend-server  enabled  active   today at 10:17 CST
+```
+
+OK, there seems no difference when everything is working. But when one service fails to start, we can see the difference - for example, when the DB can't be started (say, the port is already taken):
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:222]:
```suggestion
There seems to be no difference when everything is working. But when one service fails to start, we can see the difference. For example, here's what happens if the database can't be started (say, the port is already taken):
```

File: docs/how-to/service-dependencies.md [hunk 16]
```diff
@@ -1,11 +1,297 @@
...
+
+```{terminal}
+   :input: pebble services
+Service          Startup  Current   Since
+backend-server   enabled  inactive  -
+database         enabled  inactive  -
+frontend-server  enabled  inactive  -
+```
+
+We can see that since the database failed to start, the backend and the frontend are not started. We can further verify this in Pebble [changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md):
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:245]:
```suggestion
We can see that since the database failed to start, the backend and the frontend are not started. We can further verify this using [changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md):
```

File: docs/how-to/service-dependencies.md [hunk 17]
```diff
@@ -1,11 +1,297 @@
...
+
+Change:
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:247]: No need for this paragraph, as we already have a `:` after the paragraph above.

```suggestion
```

File: docs/how-to/service-dependencies.md [hunk 18]
```diff
@@ -1,11 +1,297 @@
...
+
+```{terminal}
+   :input: pebble changes
+ID   Status  Spawn               Ready               Summary
+1    Error   today at 10:28 CST  today at 10:28 CST  Autostart service "database" and 2 more
+```
+
+Task:
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:255]: "Task:" as a paragraph doesn't stand by itself. Perhaps just merge the two terminal blocks into one.

File: docs/how-to/service-dependencies.md [hunk 19]
```diff
@@ -1,11 +1,297 @@
...
+        return super().server_bind()
+      File "/usr/lib/python3.10/http/server.py", line 137, in server_bind
+        socketserver.TCPServer.server_bind(self)
+      File "/usr/lib/python3.10/socketserver.py", line 466, in server_bind
+        self.socket.bind(self.server_address)
+    OSError: [Errno 98] Address already in use
+2024-06-28T10:28:06+08:00 ERROR cannot start service: exited quickly with code 1
+```
+
+We can see that since the database failed to start, other services that depend on it are put in "Hold" status.
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:289]: "Hold" is actually a changes & task thing, not a service status. Maybe this:

```suggestion
We can see that since the database failed to start, other services that depend on it are not started (their service status remains "inactive").
```

File: docs/how-to/service-dependencies.md [hunk 20]
```diff
@@ -1,11 +1,297 @@
...
+
+## See more
+
+- [Pebble services command](../reference/pebble-services.md)
+- [Pebble start command](../reference/pebble-start.md)
+- [Pebble stop command](../reference/pebble-stop.md)
+- [Changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md)
+- [Layer specification](../reference/layer-specification.md)
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:297]:
```suggestion
- [`pebble services` command](../reference/pebble-services.md)
- [`pebble start` command](../reference/pebble-start.md)
- [`pebble stop` command](../reference/pebble-stop.md)
- [Changes and tasks](../reference/changes-and-tasks/changes-and-tasks.md)
- [Layer specification](../reference/layer-specification.md)
```

File: docs/how-to/service-dependencies.md [hunk 21]
```diff
@@ -1,11 +1,297 @@
 # How to manage service dependencies
 
-Pebble takes service dependencies into account when starting and stopping services. When Pebble starts a service, it also starts the services which that service depends on (configured with `required`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
+If you are using Pebble to manage your services, chances are, you've got more than one service to manage.
+
+And when orchestrating more services, things can and will get trickier, because more often than not, those services depend on each other to function together, and managing dependencies is no mean feat.
+
+Let's have a look at a concrete example of service dependencies.
+
+## My web app
```
Comment (benhoyt) [inline docs/how-to/service-dependencies.md:9]: I find "my" a bit too casual, perhaps:

```suggestion
## A simple web application
```

**Review by benhoyt — COMMENTED** [review]
Structure and approach looks good to me, but I've left a bunch of comments about style and grammar.

File: docs/how-to/service-dependencies.md [hunk 22]
```diff
@@ -1,11 +1,297 @@
 # How to manage service dependencies
 
-Pebble takes service dependencies into account when starting and stopping services. When Pebble starts a service, it also starts the services which that service depends on (configured with `required`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
+If you are using Pebble to manage your services, chances are, you've got more than one service to manage.
+
+And when orchestrating more services, things can and will get trickier, because more often than not, those services depend on each other to function together, and managing dependencies is no mean feat.
+
+Let's have a look at a concrete example of service dependencies.
```
Comment (AnneCYH) [inline docs/how-to/service-dependencies.md:7]: I suggest removing this and providing more context instead. Similar to the feedback provided about the tutorial.

E.g.

This document shows how to manage a set of services that have dependencies on each other to function properly. A simple web application is used to demonstrate the instructions in this guide.

File: docs/how-to/service-dependencies.md [hunk 23]
```diff
@@ -1,11 +1,297 @@
...
+    summary: database
+    command: python3 -m http.server 3306
+    startup: enabled
+```
+
+
+```{note}
+We use Python's http module to mock the servers.
+
+Alternatively, you may choose to define those services in a single layer. For example:
```
[the remark this answers is not in the document]
Comment (AnneCYH) [reply docs/how-to/service-dependencies.md:62]: I agree that we don't need an alternative here. If there's a reason to present single YAML vs multiple YAML files for layers, it can be a separate guide / tutorial.

I agree that if a single YAML file does the job, then the sample provided in the beginning can be shown in 1 layer instead of multiple layers.

File: docs/how-to/service-dependencies.md [hunk 24]
```diff
@@ -1,11 +1,297 @@
...
-Pebble takes service dependencies into account when starting and stopping services. When Pebble starts a service, it also starts the services which that service depends on (configured with `required`). Conversely, when stopping a service, Pebble also stops services which depend on that service.
+If you are using Pebble to manage your services, chances are, you've got more than one service to manage.
+
+And when orchestrating more services, things can and will get trickier, because more often than not, those services depend on each other to function together, and managing dependencies is no mean feat.
+
+Let's have a look at a concrete example of service dependencies.
+
+## My web app
+
+Imagine we are running a web application, which has:
```
Comment (AnneCYH) [inline docs/how-to/service-dependencies.md:11]: For how-tos I think the language should be more matter-of-fact.
https://diataxis.fr/how-to-guides/#the-language-of-how-to-guides

I suggest rewording this section to something like this.

To demonstrate the instructions in this guide, we are using a simple web
application with the following setup:

- A database listening on port 3306
- A backend server listening on port 8081, which talks to the database
- A frontend server listening on port 8080, which talks to the backend

The relationship between the frontend server, backend server and database can be simplified as:

`frontend (8080) -> backend (8081) -> database (3306)`

Comment (IronCore864) [reply docs/how-to/service-dependencies.md:11]: I merged this suggestion with Ben's.

**Comment by IronCore864** [comment]
I have resolved all comments, merged suggestions from both Anne and Ben. See the latest preview [here](https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/#).

**Comment by IronCore864** [comment]
After discussion, we decided to [merge Anne's PR](https://github.com/canonical/pebble/pull/457) first, then add back the removed sections into explanations, which is what this PR now does.

I added the sections into [Service dependencies](https://canonical-pebble--440.com.readthedocs.build/en/440/explanation/service-dependencies/) and [Service start order](https://canonical-pebble--440.com.readthedocs.build/en/440/explanation/service-start-order/), also updated links in the [how-to doc](https://canonical-pebble--440.com.readthedocs.build/en/440/how-to/service-dependencies/).

Please preview the above links and review this PR @benhoyt @AnneCYH.

**Review by AnneCYH — APPROVED** [review]

**Review by benhoyt — APPROVED** [review]
